### ISSUE TITLE
c_skip, c_out for consistency model are actually not useful?

### ISSUE THREAD

**Issue by shuminghu** [luosiallen/latent-consistency-model#82]
For the cacluation,
https://github.com/luosiallen/latent-consistency-model/blob/main/LCM_Training_Script/consistency_distillation/train_lcm_distill_sdxl_wds.py#L315-L318
```
# From LCMScheduler.get_scalings_for_boundary_condition_discrete
def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=10.0):
    c_skip = sigma_data**2 / ((timestep / 0.1) ** 2 + sigma_data**2)
    c_out = (timestep / 0.1) / ((timestep / 0.1) ** 2 + sigma_data**2) ** 0.5
    return c_skip, c_out
```

1) c_skip ~ 0, c_cout ~ 1 for any timestep !=0
here are c_skip, c_out values for timesteps 0,1,2,3 respectly
```
ipdb> scalings_for_boundary_conditions(torch.tensor([0,1,2,3]))
(tensor([1.0000e+00, 2.4938e-03, 6.2461e-04, 2.7770e-04]), tensor([0.0000, 0.9988, 0.9997, 0.9999]))
```
2) They are normalized in a weird way. c_skip + c_out**2 == 1

**Comment by jamarju** [comment]
This is somewhat explained in the Consistency Models [paper](https://arxiv.org/abs/2303.01469) (p. 3):

> For any consistency function f(·,·), we have f(x_ε, ε) = x_ε, i.e., f(·, ε) is an identity function. We call this constraint the boundary condition. All consistency models have to meet this boundary condition, as it plays a crucial role in the successful training of consistency models.

So these formulas are constructed such that for t>0, f(·,·) returns the output of the unet, but for t=0 it just returns the input image.

```
         | t=0 | t>0 |
+--------+-----+-----+
| c_skip |  1  |  ~0 |
| c_out  |  0  |  ~1 | 
+--------+-----+-----+
```

**Comment by luosiallen** [comment]
> This is somewhat explained in the Consistency Models [paper](https://arxiv.org/abs/2303.01469) (p. 3):
> 
> > For any consistency function f(·,·), we have f(x_ε, ε) = x_ε, i.e., f(·, ε) is an identity function. We call this constraint the boundary condition. All consistency models have to meet this boundary condition, as it plays a crucial role in the successful training of consistency models.
> 
> So these formulas are constructed such that for t>0, f(·,·) returns the output of the unet, but for t=0 it just returns the input image.
> 
> ```
>          | t=0 | t>0 |
> +--------+-----+-----+
> | c_skip |  1  |  ~0 |
> | c_out  |  0  |  ~1 | 
> +--------+-----+-----+
> ```

indeed.

**Comment by shuminghu** [comment]
The boundary condition is only specifed at the boundary `t=0` IIUC but didn't say c_skip ~0 , c_cout ~ 1 for t>0.
Out of all the functions that satisfiy this boundary condition, a sharp step function like what we used is counter-intuitive to me.
I was expecting something like `c_skip = exp(-at)`, and `c_out = sqrt(1-c_skip**2)`, and `a` is learnable paramater for instance. The current setup is effectively hard coding `a` to so large that `exp(-a) ~ 0 `. Such discontinuity is forced between t=0 and t=1.

![image](https://github.com/luosiallen/latent-consistency-model/assets/2934295/89f55ce2-e11b-4e12-ac35-86e0824eed84)

I was suspecting it's not useful because the probabilty of sampling t=0 to as `target_unet` input during training is small?

**Comment by jamarju** [comment]
> The boundary condition is only specifed at the boundary t=0 IIUC but didn't say c_skip ~0 , c_cout ~ 1 for t>0.

They indirectly do in the paragraph just above that:

<img width="584" alt="image" src="https://github.com/luosiallen/latent-consistency-model/assets/1566645/9c69f1b4-c3a5-4d7a-844c-6bf38d255bd6">

I think the essence of this is that you may end up trying to denoise a noiseless image, which is an impossible problem so it needs to be treated as a special case. In my opinion, using the step function above would have been easier to understand.

For visual proof, place a breakpoint [here](https://github.com/luosiallen/latent-consistency-model/blob/a9ad79587cc8bd1e404ccd1a3056a3da969b2f62/LCM_Training_Script/consistency_distillation/train_lcm_distill_sdxl_wds.py#L1310), and you'll see the attempts to denoise an image at timestep 0 every once in a while and that's the exact case where the network output gets replaced with the original image.

**Comment by shuminghu** [comment]
Thanks. 

I realized, when sampled `start_timestep` < k (DDIM step size), `timestamp` wil be [zero](https://github.com/luosiallen/latent-consistency-model/blob/a9ad79587cc8bd1e404ccd1a3056a3da969b2f62/LCM_Training_Script/consistency_distillation/train_lcm_distill_sdxl_wds.py#L1223-L1224).
This means if k=20 and `num_train_timesteps`=1000, theres 1/50 (20/1000) chance of getting input image as output for `target_unet`.  Previously i thought it was more like 1/1000.

Yes the code effectively implemented the first way of parametrization in the form of second way.

Initially, neither `unet` or `target_unet` is good, but we leverage `c_skip` to switch to noiseless image bypassing `target_unet`, we can help bootstap the the process.
0/noiseless image ---teaches---> (0, k] `unet`
(0, k] `target_unet` ---teaches----> (k, 2k] `unet`
...
So the process, albeit discrete, also works, just like Bellman update.

**Comment by shuminghu** [comment]
FYI found the formula in code similar to EDM, included in Section C of Appenedix of CM paper.
It defaults `sigma_data` to 0.5 similar to EDM but removed one `sigma_data`, which makes sense to make `c_out` close to 1.
![image](https://github.com/luosiallen/latent-consistency-model/assets/2934295/2b75bbff-a810-40ea-9ac8-fd780d975a45)

**Comment by byronyi** [comment]
If you take an actual look of $$t \in \(0, 1000\)$$ during sampling, it is way more larger than $$\sigma_{data} = 0.5$$.

So most of the time $$c_{out}$$ is just 1 and $$c_{skip}$$ is 0 (less than $$1e^{-4}$$ empirically). Just ignore this boundary condition during inference seems to make no difference to me.
